---
title: "R Notebook"
output: html_notebook
---

```{r setup, include=FALSE}
knitr::opts_knit$set(root.dir = here::here())

# library(tidyverse)
# library(vegan)
# library(ape)
# library(Biostrings)
# library(TreeDist)
```


```{r}

seq_16S <- Biostrings::readDNAStringSet("data/marker_genes/identified_variants/16S_called_variants.fasta")

V4_primers <- Biostrings::DNAStringSet(c("GTGYCAGCMGCCGCGGTAA","GGACTACNVGGGTWTCTAAT"))

V4_16S <- DECIPHER::AmplifyDNA(V4_primers,seq_16S,annealingTemp=55, P=4e-7, maxProductSize=500)

```





```{r}
seqID_to_strainID <- read_tsv("data/strain_to_seq_sample_map.tsv") %>% 
  mutate(straindID_underscores = str_replace_all(strainID, "-", "_"))

samples_to_use <- read_tsv("data/samples_to_use.tsv")

gtdb <- read_tsv("data/gtdb/gtdbtk.bac120.summary.tsv") %>% 
  filter(str_detect(classification,"Microcystis")) %>% 
  mutate(bin = user_genome)

checkM <- read_tsv("data/drep/data/checkM/checkM_outdir/results.tsv") %>% 
  mutate(bin = str_remove(`Bin Id`, ".fasta")) %>% 
  filter(bin %in% gtdb$bin)

strain_het <- checkM %>% 
  mutate(sample = str_remove(bin,"_.*"),
         strain_contam = Contamination * (`Strain heterogeneity` / 100)) %>% 
  filter(sample %in% samples_to_use$samples_to_use) %>% 
  select(sample, bin, Completeness, Contamination, `Strain heterogeneity`, strain_contam) %>% 
  group_by(sample) %>% top_n(1,Completeness) %>% 
  mutate(sequencingID = sample) %>% 
  left_join(seqID_to_strainID) %>% 
  write_tsv("results/strain_het_from_checkm.tsv")

hq_tree <- read.tree("data/gtotree/gtotree.tre")
#hq_tree <- read.tree("data/jacob_tree2.nwk")

our_strains <- hq_tree$tip.label[str_detect(hq_tree$tip.label,"LE")]

hq_tree_ours_only <- ape::keep.tip(hq_tree,tip = our_strains)

ani_tree <- read.tree("data/bin_ani_tree.nwk")
ani_tree_ours_only <- ape::keep.tip(ani_tree,tip = our_strains)

TreeDist::RobinsonFoulds(ani_tree,hq_tree)

ape::comparePhylo(ani_tree, hq_tree, plot = TRUE)
ape::comparePhylo(ani_tree_ours_only, hq_tree_ours_only, plot = TRUE)

variant_tree_files <- list.files("data/marker_genes/identified_variants",pattern = ".tre",full.names = TRUE)

variant_trees <- lapply(variant_tree_files,read.tree)

tree_names <- character()

for (i in 1:length(variant_trees)){
  tip_labels <- data.frame(old.label = variant_trees[[i]]$tip.label) %>% 
     mutate(gene = str_remove(old.label, "_.*"),
            sequencingID = str_remove(old.label,"_consensus") %>% str_remove(".*_")) %>% 
     left_join(seqID_to_strainID)
  
  tree_names <- c(tree_names, tip_labels$gene[1])
  
  print(tip_labels$gene[1])
   
  variant_trees[[i]]$tip.label <- tip_labels$straindID_underscores
  
  variant_trees[[i]]<- keep.tip(variant_trees[[i]],our_strains)
}

names(variant_trees) <- tree_names


TreeDist::RobinsonFoulds(hq_tree_ours_only %>% keep.tip(variant_trees[["16S"]]$tip.label),variant_trees[["16S"]])

ape::comparePhylo(hq_tree_ours_only %>% keep.tip(variant_trees[["16S"]]$tip.label),variant_trees[["pgi"]], plot = TRUE)



t1 <- hq_tree_ours_only %>% keep.tip(variant_trees[["pgi"]]$tip.label) %>% phytools::midpoint.root() %>% ape::chronoMPL() %>% as.dendrogram()
t2 <- variant_trees[["pgi"]] %>% phytools::midpoint.root() %>% ape::chronoMPL()   %>% as.dendrogram()

assoc <- cbind(t1$tip.label, t1$tip.label)
#cophyloplot(t1, t2, assoc, length.line = 4, space = 28, gap = 3)


dndlist <- dendextend::dendlist(t1, t2)
dendextend::tanglegram(dndlist, fast = TRUE, margin_inner = 1.8, lab.cex = 0.3, lwd = 
0.5, edge.lwd = 0.5, type = "r",)


compare_to_hq_tree <- function(tree){
  TreeDist::RobinsonFoulds(hq_tree_ours_only %>% keep.tip(tree$tip.label),tree)
}

ani_dist_to_multigene_tree <- TreeDist::RobinsonFoulds(ani_tree_ours_only,hq_tree_ours_only)

tree_distances <- sapply(variant_trees, compare_to_hq_tree) %>% 
  data.frame(dist_to_hq_tree = ., gene = names(.)) %>% 
  bind_rows(data.frame(dist_to_hq_tree = ani_dist_to_multigene_tree, gene = "ani"))
```



Phylomark results
```{r}
phylomark_results <- read_tsv("data/phylomark/400bp/results.txt") 

phylomark_top100 <- phylomark_results %>% 
  top_n(100,-RF)

subset_tree <- function(tree){
  trimmed_tree <- ape::keep.tip(tree,tip = our_strains)
  return(trimmed_tree)
}

dist_to_hq_tree <- function(tree){
  dist_to_multigene_tree <- TreeDist::RobinsonFoulds(tree,hq_tree_ours_only)
  return(dist_to_multigene_tree)
}


phylomark_trees <- read_tsv("data/phylomark/400bp/all_trees.txt", col_names = c("sequence", "tree")) %>% 
  filter(sequence %in% phylomark_top100$sequence) %>% 
  left_join(phylomark_top100) %>% 
  rowwise() %>% 
  mutate(tree_formatted = list(ape::read.tree(text = tree)),
         phylomark_our_strains = list(subset_tree(tree_formatted)),
         phylomark_RF_our_strains = dist_to_hq_tree(phylomark_our_strains),
         type = "phylomark") %>% 
  ungroup() %>% 
  mutate(rounded_seq = plyr::round_any(sequence,100)) %>% 
  group_by(rounded_seq) %>% 
  arrange(RF) %>% 
  mutate(order = row_number()) %>% 
  filter(order == 1) %>% 
  arrange(phylomark_RF_our_strains) %>% 
  mutate(order = row_number()) %>% 
  filter(order == 1)

phylomark_seqs <- Biostrings::readDNAStringSet("data/phylomark/400bp/query_sequences.fasta") %>% 
  data.frame(seq = ., header = names(.))
  

phylomark_trees %>% 
  ggplot(aes(RF, phylomark_RF_our_strains, color = `#polymorphisms`, label = sequence)) +
  geom_point() +
  geom_label() +
  theme_bw() +
  labs(x = "RF distance to multi-gene tree for all microcystis",
       y = "RF distance to multi-gene tree for WLE strains") +
  scale_color_viridis_c()

phylomark_and_genes <- phylomark_trees %>% 
  mutate(marker = as.character(sequence)) %>% 
  select(all_microcystis_RF = "RF", dist_to_hq_tree = "phylomark_RF_our_strains", marker, type) %>% 
  full_join(tree_distances %>% 
              mutate(type = "existing_marker") %>% 
              dplyr::rename(marker = "gene"))


phylomark_genes <- data.frame(marker = c("570950", "35920", "660160", "670124", "264397"),
                              gene = c("accD", "GspE/PulE", "nrdR", "rpoC2", "Rne/Rng"))

original_genes <- c("16S", "16SV2", "16SV4", "pgi", "recA", "glnA", "itsC")

phylomark_and_genes %>% 
  left_join(phylomark_genes) %>% 
  mutate(marker = if_else(!is.na(gene), paste0(gene,"_",marker),marker)) %>% 
  mutate(type = if_else(marker == "ani", "ANI", type)) %>% 
  ggplot(aes(dist_to_hq_tree, reorder(marker,-dist_to_hq_tree), fill = type)) +
  geom_bar(stat = "identity") +
  scale_fill_viridis_d(alpha = 0.8) +
  theme_bw() +
  labs(y = "Marker / gene", "RF distance to multi-gene Microcystis tree") 
ggsave("results/phylo_markers.png", height = 3, width = 3, dpi = 600, scale = 2)

```



Check for heterogeneity in read mappings
```{r}

alignments <- list.files("data/marker_genes/identified_variants","_called_variants.afa",full.names = TRUE)

read_alignments <- function(path){
  
  gene_name <- str_remove(path,".*/") %>% str_remove("_called_variants.afa")
  
  alignment <- Biostrings::readDNAMultipleAlignment(path) %>% 
    consensusMatrix() %>% 
    as.data.frame()

  div <- diversity(alignment,MARGIN = 2) %>% 
    data.frame(position = names(.), diversity = .) %>% 
    mutate(position = as.numeric(str_remove(position,"V")),
           gene = gene_name) 

  return(div)
}


test <- read_alignments(alignments[1])

alignments_div <- map_df(alignments,read_alignments)

gene_div <- alignments_div %>% 
  group_by(gene) %>% 
  summarise(gene_length = max(position),
         div_sum = sum(diversity),
         norm_div = div_sum/gene_length)

alignments_div %>% 
  ggplot(aes(position,diversity)) + 
  geom_bar(stat = "identity") +
  facet_grid(gene ~. ,scales = "free_x")
```



```{r}


param <- Rsamtools::ScanBamParam(what = "seq")

gal  <- GenomicAlignments:::readGAlignments("data/marker_genes/ND-100/16S_mapped_to_consensus.bam", param=param)
qseq <- mcols(gal)$seq

qseq_on_ref <- GenomicAlignments::sequenceLayer(qseq, GenomicAlignments::cigar(gal),
                                    from="query", to="reference")



qseq_on_ref_by_chrom <- splitAsList(qseq_on_ref, seqnames(gal))
qseq_pos_by_chrom <- splitAsList(start(gal), seqnames(gal))

cm_by_chrom <- lapply(names(qseq_pos_by_chrom),
   function(seqname)
       consensusMatrix(qseq_on_ref_by_chrom[[seqname]],
                       as.prob=TRUE,
                       shift=qseq_pos_by_chrom[[seqname]]-1,
                       width=seqlengths(gal)[[seqname]]))
names(cm_by_chrom) <- names(qseq_pos_by_chrom)

test <- data.frame(cm_by_chrom) %>% 
  rownames_to_column("base") %>% 
  pivot_longer(-base, names_to = "position", values_to = "count") %>% 
  mutate(position = as.character(str_remove(position, ".*\\."))) 


test %>% 
  ggplot(aes(position, count, color = base)) + 
  geom_point()

test %>% 
  ggplot(aes(count)) +
  geom_histogram() + scale_y_log10()

```


```{r}

no_index <- function(x){
  !str_detect(x, "bai")
}


bams_to_consensus <- list.files("data/marker_genes","_mapped_to_consensus.bam",recursive = TRUE, full.names = TRUE) %>% 
  Filter(x = .,f = no_index )

bam_path <- bams_to_consensus[1]

bam_to_consensusMatrix <- function(bam_path){
  sample <- str_remove(bam_path,".*marker_genes/") %>% str_remove("/.*")
  gene <- str_remove(bam_path,".*/") %>% str_remove("_mapped_to_consensus.bam")
  
  param <- Rsamtools::ScanBamParam(what = "seq")
  gal  <- GenomicAlignments:::readGAlignments(bam_path, param=param)
  qseq <- mcols(gal)$seq
  
  qseq_on_ref <- GenomicAlignments::sequenceLayer(qseq, GenomicAlignments::cigar(gal),
                                      from="query", to="reference")
  
  qseq_on_ref_by_chrom <- splitAsList(qseq_on_ref, seqnames(gal))
  qseq_pos_by_chrom <- splitAsList(start(gal), seqnames(gal))
  
  cm_by_chrom <- lapply(names(qseq_pos_by_chrom),
     function(seqname)
         consensusMatrix(qseq_on_ref_by_chrom[[seqname]],
                         as.prob=FALSE,
                         shift=qseq_pos_by_chrom[[seqname]]-1,
                         width=seqlengths(gal)[[seqname]]))
  names(cm_by_chrom) <- names(qseq_pos_by_chrom)
  
  df <- data.frame(cm_by_chrom) %>% 
    rownames_to_column("base") %>% 
    pivot_longer(-base, names_to = "position", values_to = "count") %>% 
    mutate(position = as.character(str_remove(position, ".*\\.")),
           gene = gene,
           sample = sample) 
  
  return(df)
}



test <- map_df(bams_to_consensus,bam_to_consensusMatrix) %>% 
  group_by(position,gene,sample) %>% 
  mutate(site_coverage = sum(count)) %>% 
  #filter(sample %in% samples_to_use$samples_to_use) %>%
  ungroup() %>% 
  group_by(gene, position, sample) %>% 
  arrange(desc(count)) %>% 
  mutate(rank = row_number(),
         percent = count / site_coverage * 100)


diversity_matrix <- test %>% 
  select(sample, gene, base, position, count) %>% 
  pivot_wider(names_from = base, values_from = count)


diversity <- test %>% 
  group_by(sample, gene, position) %>% 
  summarise(shannon = vegan::diversity(count))

diversity_summary <- diversity %>% 
  ungroup() %>% group_by(sample) %>% 
  summarise(mean_shannon = mean(shannon)) %>%
  left_join(seqID_to_strainID %>% dplyr::rename(sample = "sequencingID")) %>% 
  write_tsv("results/strain_het_by_marker_gene_mean_shannon_diversity.tsv")

test_simplified <- test %>% 
  filter(count > 2) %>% 
  group_by(gene,position,sample) %>% 
  mutate(n_snp_types = n()) %>% 
  filter(n_snp_types > 1)

test %>% filter(rank == 1) %>% 
  ggplot(aes(sample,percent, color = gene, group =sample)) +
  #geom_density_2d()
  geom_jitter() +
  geom_boxplot() +
  scale_y_log10()
  #geom_boxplot() 
  #geom_point() +
  #ggridges::geom_density_ridges(panel_scaling = TRUE) +
  #scale_x_continuous(expand = c(0, 0))


test %>% filter(rank ==1) %>% 
  ggplot(aes(position,percent,color = gene)) +
  geom_point() +
  facet_wrap(sample ~ .)

(snps_plot <- test_simplified %>% filter(rank == 1, count > 2) %>% 
  mutate(sequencingID = sample) %>% 
  left_join(seqID_to_strainID) %>% 
  filter(sample %in% samples_to_use$samples_to_use) %>% 
  filter(strainID %in% strain_het$strainID) %>% 
  filter(strainID %in% snps_bar$data$strainID) %>%
  ggplot(aes(position,percent,color = gene)) +
  geom_point() +
  facet_grid(~strainID) +
    theme(legend.position = "none")
)


(snps_plot <- test_simplified %>% filter(rank == 1, count > 2) %>% 
  mutate(sequencingID = sample) %>% 
  left_join(seqID_to_strainID) %>% 
  filter(sample %in% samples_to_use$samples_to_use) %>% 
  filter(strainID %in% strain_het$strainID) %>% 
  filter(strainID %in% snps_bar$data$strainID) %>%
  ggplot(aes(position,100 - percent,color = gene)) +
  geom_point() +
    labs(y = "% non-primary base") +
  facet_grid(~strainID) 
    #theme(legend.position = "none")
)

snp_summary <- test_simplified %>% 
  group_by(sample) %>% 
  filter(rank ==1,
         percent < 100, 
         count > 2,
         sample %in% samples_to_use$samples_to_use) %>% 
  mutate(n_snps = n(),
         mean_primary_snp_percent = mean(percent),
         SNP_summary = (100 - mean_primary_snp_percent) / 100 * n_snps) %>% 
  select(sample, n_snps, mean_primary_snp_percent, SNP_summary) %>% 
  mutate(sequencingID = sample) %>% 
  left_join(seqID_to_strainID) %>% 
  distinct() 
    
snp_summary %>% ggplot(aes(n_snps,SNP_summary, color = strainID, label = strainID)) +
    geom_point() +
    geom_label() +
    scale_y_log10()

(SNP_summary_plot <- snp_summary %>% 
  filter(strainID %in% strain_het$strainID) %>% 
  filter(strainID %in% snps_bar$data$strainID) %>%
  ggplot(aes(strainID,1, fill = SNP_summary)) +
    geom_tile() + 
    labs(x = NULL) +
    theme(legend.position = "none") +
  scale_fill_viridis_c()
)

seconds_simplified <- test_simplified %>% 
  filter(rank > 1 & count > 0) 

seconds_simplified %>% 
  filter(count > 2) %>% 
  ggplot(aes(percent,sample, color = gene)) +
  #geom_density_2d()
  geom_jitter() +
  #geom_point() +
  #ggridges::geom_density_ridges(panel_scaling = TRUE) +
  #scale_x_continuous(expand = c(0, 0))
  scale_x_log10()
  #scale_x_log10() + 
  #scale_y_log10() 
  #facet_wrap(sample ~ .)


snps_w_2plus_reads <- seconds_simplified %>% 
  ungroup() %>% 
  filter(count > 2) %>% 
  group_by(sample,gene) %>% 
  summarise(num_snps = n()) %>% 
  filter(sample %in% samples_to_use$samples_to_use) %>% 
  mutate(sequencingID = sample) %>% 
  left_join(seqID_to_strainID)

(snps_bar <- snps_w_2plus_reads %>% 
  filter(strainID %in% strain_het$strainID) %>% 
  ggplot(aes(gene, num_snps, fill = gene)) +
  #geom_jitter() +
  geom_bar(stat = "identity", position = "dodge") +
  facet_grid( ~strainID) +
  labs(x = NULL, y = "Number of SNPs") +
  theme(axis.text.x = element_blank(), axis.ticks.x = element_blank()
    #axis.text.x = element_text(angle = -45, hjust = 0, vjust = 0)
    )
)
  
(checkM_strains <- strain_het %>% 
    filter(strainID %in% snps_bar$data$strainID) %>% 
  ggplot(aes(strainID, y = 1, fill = strain_contam)) +
  geom_tile(color = "white") +
  theme(axis.text.x = element_text(angle = -45, vjust = 0, hjust = 0)) + 
  scale_fill_viridis_c(trans = "log") +
  labs(x = NULL,
       y=NULL,
       fill = "checkM \nstrain contam."
       #y= "checkM \nstrain contam."
       ) +
  theme(axis.text.y = element_blank(),
        axis.ticks.y = element_blank(),
        #legend.position="none"
        )
)

(shannon_heatmap <- diversity_summary %>% 
  mutate(sequencingID = sample) %>% 
  left_join(seqID_to_strainID) %>% 
  filter(strainID %in% snps_bar$data$strainID) %>% 
  ggplot(aes(strainID, 1, fill = mean_shannon)) +
  scale_fill_viridis_c() +
    geom_tile()
)

library(patchwork)

snps_plot  / SNP_summary_plot / shannon_heatmap / checkM_strains + plot_layout(heights = c(0.4, 0.1, 0.1 , 0.1))
snps_plot / checkM_strains + plot_layout(heights = c(0.8, 0.2))
ggsave("results/strain_heterogeneity.png", width = 10, height = 4, scale =2, dpi= 600)

```





bamfile <- system.file("extdata", "ex1.bam",
package="Rsamtools")
       param <- ScanBamParam(what="seq")
       gal <- readGAlignmentsFromBam(bamfile, param=param)
       qseq <- mcols(gal)$seq  # the query sequences

(2) Use sequenceLayer() to "lay" the query sequences on the reference
     space. This will remove the parts from the query sequences that
     correspond to insertions and soft clipping, and it will fill them
     with - where deletions and/or skipped regions occurred:

       qseq_on_ref <- sequenceLayer(qseq, cigar(gal),
                                    from="query", to="reference")

(3) Compute 1 consensus matrix per chromosome:

       qseq_on_ref_by_chrom <- splitAsList(qseq_on_ref, seqnames(gal))
       qseq_pos_by_chrom <- splitAsList(start(gal), seqnames(gal))

       cm_by_chrom <- lapply(names(qseq_pos_by_chrom),
           function(seqname)
               consensusMatrix(qseq_on_ref_by_chrom[[seqname]],
                               as.prob=TRUE,
                               shift=qseq_pos_by_chrom[[seqname]]-1,
                               width=seqlengths(gal)[[seqname]]))
       names(cm_by_chrom) <- names(qseq_pos_by_chrom)





Make primers
```{r}
library(DECIPHER)

fas <- "data/marker_genes/identified_variants/pgi_called_variants.fasta"

dbConn <- dbConnect(SQLite(), ":memory:")
N <- Seqs2DB(fas, "FASTA", dbConn, "")

desc <- dbGetQuery(dbConn, "select description from Seqs")$description
Add2DB(data.frame(identifier=desc, stringsAsFactors=FALSE), dbConn)


# Designing primers for sequencing experiments:
TYPE <- "sequence"
MIN_SIZE <- 275 # base pairs
MAX_SIZE <- 450
RESOLUTION <- 5 # k-mer signature
LEVELS <- 5 # max number of each k-mer
ENZYMES <- NULL # required for sequencing

primers <- DesignSignatures(dbConn,tblName = "Seqs",
    type=TYPE,
    minProductSize=MIN_SIZE,
    maxProductSize=MAX_SIZE,
    resolution=RESOLUTION,
    levels=LEVELS,
    enzymes=ENZYMES)

```





